Refactor pact mock server network calling code and upgrade to swift 5

surpher opened the pull request:
This pull request resolves part of #48. 

- Updates README to include brew tap
- Uses brew tap to install `pact-ruby-standalone` as PactMockService on CI
- Introduces a Network Layer to use when making network calls
- Removes Alamofire dependency from /Sources and /Tests
- Rewrites tests leaving the same interface as before

Commit:
Gets the project built

Update base proejct to Swift5 and targets to use undefined swift version (inherit from base project)

Commit:
Update travisCI config to run Xcode10.2 with swift 5

Commit:
Fixes the enum empty value error

Commit:
Update README [skip-ci]

Install pact-mock-service using homebrew

Commit:
Sepcify Swift version 5 for tvOS and macOS

surpher committed:
Remove unused import

surpher committed:
Update Swiftlint rules

- set `line_length` warning and error
- exclude keywords for HTTP Methods from identifier_name

surpher committed:
Update Swiftlint rules

- set `line_length` warning and error
- exclude keywords for HTTP Methods from identifier_name

surpher committed:
Add Separated NetworkLayer

Removes dependency on Alamofire

surpher committed:
Add Separated NetworkLayer

Removes dependency on Alamofire

surpher committed:
Removes Alamofire

surpher committed:
Removes Alamofire

surpher committed:
Remove Alamofire from Xcode Project and dep managers

surpher committed:
Remove Alamofire from Xcode Project and dep managers

surpher committed:
Add Network Layer files to list of Compiled Sources

surpher committed:
Add Network Layer files to list of Compiled Sources

surpher committed:
Clean up Project file

surpher committed:
Clean up Project file

surpher committed:
Update CI step to run mock from brew tap

surpher committed:
Update CI step to run mock from brew tap

andrewspinks commented on .swiftlint.yml:
```diff
@@ -34,4 +37,9 @@ identifier_name:
     - min
     - max
+    - key
+    - GET
```
Some of these are duplicated GET, PUT

surpher replied:
🤦‍♂Yep, I'm blind.

surpher committed:
Merge branch 'master' of github.com:DiUS/pact-consumer-swift

# Conflicts:
#	Cartfile.resolved
#	Sources/PactVerificationService.swift

surpher committed:
Merge branch 'maintenance/remove_alamofire_dep' of github.com:surpher/pact-consumer-swift into maintenance/remove_alamofire_dep

surpher committed:
Add flag to enable networkLogging

Initialising `PactVerificationService` wtih `networkLogging = true` will display calls and responses to the Pact Mock Service to help with debugging of url requests and data passed between the client and provider.

surpher committed:
Update build script to exit at non-zero status from a cmd

surpher committed:
Update CI to use pact-foundation tap for pact-ruby-standalone

surpher committed:
Resolves outdated SwiftPM packages

surpher committed:
Documentation style comments for public methods

surpher committed:
Update README.md

surpher committed:
Remove redundant hardcoded header value

andrewspinks commented on Sources/MockAPI.swift:
```diff
@@ -0,0 +1,75 @@
+
+}
+
+extension MockAPI: EndPointType {
```
Maybe call this MockServiceAPI or PactMockServiceAPI? When I read this initially, I thought it was just for unit tests (i.e. it was a mock class).

surpher replied:
Agreed. Renamed it to `PactMockServiceAPI`

andrewspinks commented on Sources/Result.swift:
```diff
@@ -0,0 +1,6 @@
+import Foundation
+
+enum Result<String> {
```
Maybe should be using this now? https://developer.apple.com/documentation/swift/result

surpher replied:
I'll have a look. Might be tempted to do it in a separate PR.

surpher replied:
Sorted it out in this PR

surpher committed:
Set code indentation to 2 spaces

surpher committed:
Changed indentation to 2 spaces

surpher committed:
Remove trailing empty lines

andrewspinks commented on Sources/HTTPTask.swift:
```diff
@@ -0,0 +1,14 @@
+  case requestParameters(bodyParameters: Parameters?,
+    urlParameters: Parameters?)
+
+  case requestParametersAndHeaders(bodyParameters: Parameters?,
```
Whats the reason for having both requestParameters and requestParametersAndHeaders when the headers is nullable anyway?

andrewspinks commented on Sources/Interaction.swift:
```diff
@@ -60,6 +114,7 @@ open class Interaction: NSObject {
 
+  ///
   @objc
   open func payload() -> [String: Any] {
```
I wonder if this should have been `internal`. Maybe leave for another day.

surpher replied:
I have personally never used `.payload()` but I see there are tests in the framework that use it quite a bit. I'd leave it for a separate PR if it an eyesore.

andrewspinks commented on Sources/JSONParameterEncoder.swift:
```diff
@@ -0,0 +1,16 @@
+      let jsonAsData = try JSONSerialization.data(withJSONObject: parameters,
+                                                  options: .prettyPrinted)
+      urlRequest.httpBody = jsonAsData
+      if urlRequest.value(forHTTPHeaderField: "Content-Type") == nil {
```
Should be fine to just set the content type regardless.

surpher replied:
I'd personally avoid overwriting regardless, but I'm not married to it either. So I'll take the subtle suggestion.

andrewspinks commented on Sources/MockService.swift:
```diff
@@ -10,76 +10,166 @@ open class MockService: NSObject {
+  /// Use this method in the `Arrange` step of your Pact test.
+  ///
+  ///     myMockService.given("a user exists")
+  ///                  .uponReceiving("a request for users")
```
This is the entrypoint if not using a provider state. i.e.

```myMockService.uponReceiving("a request for users")```

andrewspinks commented on Sources/MockService.swift:
```diff
@@ -10,76 +10,166 @@ open class MockService: NSObject {
+  /// Make sure you call `testComplete()` after your `Assert` step in your test
+  ///
+  /// - Parameter file: The file in which your test resides
+  /// - Parameter line: The line number of your test
```
Probably don't want to document the `file` and `line` parameters. They are filled in by the compiler with the file and line of the calling method, which we then use in failWithLocation to cause xcode to highlight the test calling code as in error (rather than somewhere in this library). If someone supplies the parameters then xcode will highlight the error as coming from a different place in the code base.

andrewspinks commented on Sources/MockService.swift:
```diff
@@ -10,76 +10,166 @@ open class MockService: NSObject {
+  /// Use this method in the `Act` step of your Pact test.
+  /// (eg. Testing your `serviceClientUnderTest!.getUsers(...)` method)
+  ///
+  ///     myMockService!.run(line: 40, timeout: 10) { (testComplete) -> Void in
```
Remove line from the example. Don't want to encourage its use.

andrewspinks commented on Sources/MockService.swift:
```diff
@@ -10,76 +10,166 @@ open class MockService: NSObject {
   }
 
+  ///
+  /// Runs the provided test function by specifying timeout in seconds
```
This is method is only her for objective-c compatibility so the example should be:

```
[self.mockService run:^(CompleteBlock testComplete) {
     [self. serviceClientUnderTest getUsers]
      testComplete();
} withTimeout: 10];
```

andrewspinks commented on Sources/MockService.swift:
```diff
@@ -10,76 +10,166 @@ open class MockService: NSObject {
+  ///
+  /// Use this method in the `Act` step of your Pact test.
+  /// (eg. Testing your `serviceClientUnderTest!.getUsers(...)` method)
+  ///
```
This is method is only her for objective-c compatibility so the example should be:

```
[self.mockService run:^(CompleteBlock testComplete) {
     [self. serviceClientUnderTest getUsers]
      testComplete();
}];
```

andrewspinks commented on Sources/Router.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+  }
+
+  fileprivate func addAdditionalHeaders(_ additionalHeaders: HTTPHeaders?, request: inout URLRequest) {
```
Why are the headers being called `additionalHeaders` and not just `headers`?

andrewspinks commented on Tests/PactConsumerSwiftTests/AnimalServiceClient.swift:
```diff
@@ -1,164 +1,268 @@
+        (200..<400).contains(response.statusCode) ? success() : error(response.statusCode)
       } else {
-        success()
+        error(500)
```
Shouldn't this be returning the status code from the service?

surpher replied:
Is/Was a _bad_ client implementation for the test purpose so I personally wasn't too fussed at what it does

andrewspinks commented on Tests/PactConsumerSwiftTests/AnimalServiceClient.swift:
```diff
@@ -1,164 +1,268 @@
+        (200..<400).contains(response.statusCode) ? success() : error(response.statusCode)
       } else {
-        success()
+        error(500)
```
Shouldn't this be returning the status code from the service?

surpher replied:
Is/Was a _bad_ client implementation for the test purpose so I personally wasn't too fussed at what it does

andrewspinks reviewed (commented):
Can we rename this pull request to something like: `Refactor pact mock server network calling code and upgrade to swift 5`

Do we also need to change the version in `.swift-version`

surpher committed:
Remove duplicated swiftlint identifer_name rules

surpher committed:
Rename Endpoint API to PactMockServiceAPI

surpher committed:
Set urlRequest header values regardless

surpher committed:
Remove redundant case for setting nullable parameters

surpher committed:
Update documentation comments

surpher committed:
Rename `additionalHeaders` to `headers`

surpher committed:
Changed return type in demo AnimalServiceClientApi

Returning an error object on failed URLResponse. When response data is missing, return an error object instead of 500.

surpher committed:
Using Swift5 Response<String, Error>

surpher commented:
> Can we rename this pull request to something like: `Refactor pact mock server network calling code and upgrade to swift 5`
> 
> Do we also need to change the version in `.swift-version`

I have no idea when (and I didn't go back in git log searching for it) and why this `.swift-version` got introduced in the first place tbh.

But apparently it is now deprecated and Pods only use the attribute within the podspec now... I vote it gets removed separately.
